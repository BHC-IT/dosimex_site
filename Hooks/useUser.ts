import { useState, useEffect } from 'react';

export interface IUser {
	username: string,
	password: string,
}


const useUser = () => {

	const [token, setToken] = useState<string | null>(null)

	useEffect(() => {
		setToken(localStorage.getItem('access_token'))
	}, []);

	return token;
}

export default useUser;
