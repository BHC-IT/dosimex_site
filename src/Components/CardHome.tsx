import * as React from 'react';
import Link from 'next/link';
import Image from 'next/image';
import Radium from 'radium';
import * as CSS from 'csstype';

export interface IStyles {
	button: CSS.Properties,
	subtitle: CSS.Properties,
	card: CSS.Properties,
	image: CSS.Properties,
}

interface IProps {
	icon: string,
	title: string,
	content: string,
	route: string,
}

const CardHome = (props: IProps) => {

	return (
		<div style={styles.card}>
			<div style={styles.image}>
				<Image
					src={props.icon}
					alt="icône"
					width={158}
					height={158}
				/>
			</div>
			<h4 style={{marginBottom: "0"}}>{props.title}</h4>
			<p style={styles.subtitle}>{props.content}</p>
			<button style={styles.button}>
				<Link href={`/${props.route}`} replace><span>→</span></Link>
			</button>
		</div>
	);

}

export default Radium(CardHome);

export const styles =  {
	button: {
		width: "40px",
		height: "40px",
		display: "flex",
		alignItems: "center",
		justifyContent: "center",
		fontSize: "3rem",
		padding: "20px",
		marginLeft: "auto",
		marginRight: "auto",
		marginTop: "2vh",
		marginBottom: "4vh",
		backgroundColor: "var(--main)",
		borderRadius: "50%",
		color: "white",
		textTransform: "uppercase",
		transition: "all 0.3s ease 0s",
		':hover': {
			color: "white",
			transform: "translateY(-2px)",
			boxShadow: "0px 5px 5px rgba(0, 0, 0, 0.1)",
		},
	},
	card: {
		borderRadius: "10px",
		boxShadow: "0px 5px 5px rgba(0, 0, 0, 0.1)",
		textAlign: "center",
		width: "28%",
	},
	subtitle: {
		color: "var(--grey)",
		fontSize: "1.8rem",
		marginBottom: "4vh",
		marginLeft: "2%",
		marginRight: "2%",
	},
	image: {
		width: "20%",
		height: "20%",
		marginLeft: "auto",
		marginRight: "auto",
	},
}