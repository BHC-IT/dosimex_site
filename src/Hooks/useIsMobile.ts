import * as React from 'react'

import { isMobile, isTablet } from 'react-device-detect'

export const useIsMobile = <T>(styles: (m: boolean) => T): T | null => {
	// @ts-ignore - React hooks compatibility with React 18
	const [style, setStyle] = React.useState<T | null>(styles(isMobile))

	// @ts-ignore - React hooks compatibility with React 18
	React.useEffect(() => {
		setStyle(null)
		setTimeout(() => setStyle(styles(isMobile)), 10)
	}, [])

	return style
}

export const useMobile = (): boolean | null => {
	// @ts-ignore - React hooks compatibility with React 18
	const [mobile, setMobile] = React.useState<boolean | null>(null)

	// @ts-ignore - React hooks compatibility with React 18
	React.useEffect(() => {
		setTimeout(() => setMobile(isMobile), 10)
	}, [])

	return mobile
}

export const useTablet = (): boolean | null => {
	// @ts-ignore - React hooks compatibility with React 18
	const [tablet, setTablet] = React.useState<boolean | null>(null)

	// @ts-ignore - React hooks compatibility with React 18
	React.useEffect(() => {
		setTimeout(() => setTablet(isTablet), 10)
	}, [])

	return tablet
}

export const useDeviceType = (): { isMobile: boolean | null; isTablet: boolean | null } => {
	const mobile = useMobile()
	const tablet = useTablet()

	return { isMobile: mobile, isTablet: tablet }
}
