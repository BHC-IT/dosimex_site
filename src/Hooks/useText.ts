import { useRouter } from 'next/router';

import ILang from '../lang/interface';

import fr from '../lang/fr';
import en from '../lang/en';
import debug from '../lang/debug';

export const text : {[$:string]: ILang} = {fr, en, debug};

export const useText = (page: string) : any => {
	const route = useRouter();

	if (route.locale === 'debug')
		return text[route.locale][page];
	else if (route?.locale)
		return text[route.locale.slice(0,2)]?.[page];
	return null
}
