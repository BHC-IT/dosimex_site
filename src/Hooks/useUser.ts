import { useState, useEffect } from 'react';


export interface IUser {
	username: string,
	password: string,
}

const remove = (setToken: Function) => {
	localStorage.removeItem('access_token')
	localStorage.removeItem('exp_token')
	setToken(null)
}

const useUser = () => {

	const [token, setToken] = useState<string | null>(null)

	useEffect(() => {
		setToken(localStorage.getItem('access_token'))
		const exp = Number(localStorage.getItem('exp_token'))

		if (!exp) {
			remove(setToken)
			return;
		}
		if (exp < Date.now())
			remove(setToken)
		else
			setTimeout(() => remove(setToken), exp - Date.now())
	}, []);

	return token;
}

export default useUser;
