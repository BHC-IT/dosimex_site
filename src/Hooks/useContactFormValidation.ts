import emailjs from '@emailjs/browser'
import { useState, useCallback } from 'react'
import { isPossiblePhoneNumber } from 'react-phone-number-input'
import { toast } from 'react-toastify'

import { IValidator } from '../Components/Input'

// Email validation regex
const EMAIL_REGEX =
	/^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/

// Validation functions
export const isInputValid = (value: string): boolean => value.trim() !== ''
export const isEmailValid = (value: string): boolean => EMAIL_REGEX.test(value)
export const isPhoneValidNumber = (phone: string | undefined): boolean =>
	!phone || isPossiblePhoneNumber(phone)

// Validator factory functions
export const createRequiredValidator = (errorMessage: string): IValidator => ({
	validationFunction: value => isInputValid(value),
	errorMessage,
})

export const createEmailValidator = (errorMessage: string): IValidator => ({
	validationFunction: value => isEmailValid(value),
	errorMessage,
})

export interface IContactFormValidation {
	// Form field states
	name: string | null
	email: string | null
	phone: string | undefined
	enterprise: string | null
	address: string | null
	subject: string | null
	message: string | null

	// Validation states
	nameValid: boolean
	phoneError: string | null
	isLoading: boolean

	// Form handlers
	setName: (value: string) => void
	setEmail: (value: string) => void
	setPhone: (value: string | undefined) => void
	setEnterprise: (value: string) => void
	setAddress: (value: string) => void
	setSubject: (value: string) => void
	setMessage: (value: string) => void
	setNameValid: (valid: boolean) => void
	setIsLoading: (loading: boolean) => void

	// Validation methods
	validatePhone: () => void
	validateForm: () => boolean
	resetForm: () => void
	handleSubmit: (e: React.FormEvent) => void

	// Validator configurations
	getNameValidators: (errorMessage: string) => IValidator[]
	getEmailValidators: (errorEmail: string[]) => IValidator[]
	getAddressValidators: (errorMessage: string) => IValidator[]
}

// Validation helper functions
const validateName = (name: string | null): boolean => {
	return name !== null && name.trim() !== ''
}

const validateEmail = (email: string | null): boolean => {
	if (!email || email.trim() === '') return false
	const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/
	return emailRegex.test(email)
}

const validateMessage = (message: string | null): boolean => {
	return message !== null && message.trim() !== ''
}

// Custom hook for form state management
const useFormState = () => {
	const [name, setName] = useState<string | null>(null)
	const [email, setEmail] = useState<string | null>(null)
	const [phone, setPhone] = useState<string | undefined>(undefined)
	const [enterprise, setEnterprise] = useState<string | null>(null)
	const [address, setAddress] = useState<string | null>(null)
	const [subject, setSubject] = useState<string | null>(null)
	const [message, setMessage] = useState<string | null>(null)
	const [nameValid, setNameValid] = useState(false)
	const [isLoading, setIsLoading] = useState(false)

	return {
		name, setName,
		email, setEmail,
		phone, setPhone,
		enterprise, setEnterprise,
		address, setAddress,
		subject, setSubject,
		message, setMessage,
		nameValid, setNameValid,
		isLoading, setIsLoading,
	}
}

// Form validation logic
const useFormValidation = (nameValid: boolean, phone: string | undefined) => {
	const checkInputs = useCallback((): boolean => {
		return nameValid && (phone ? isPossiblePhoneNumber(phone) : true)
	}, [nameValid, phone])

	const validateForm = useCallback(() => checkInputs(), [checkInputs])

	const validatePhone = useCallback(() => {
		// Phone validation is done in checkInputs
	}, [])

	return { checkInputs, validateForm, validatePhone }
}

// Form reset functionality
const useFormReset = (
	setName: (value: string | null) => void,
	setEmail: (value: string | null) => void,
	setPhone: (value: string | undefined) => void,
	setEnterprise: (value: string | null) => void,
	setAddress: (value: string | null) => void,
	setSubject: (value: string | null) => void,
	setMessage: (value: string | null) => void,
	setNameValid: (valid: boolean) => void,
	setIsLoading: (loading: boolean) => void,
) => {
	const resetForm = useCallback(() => {
		setName(null)
		setEmail(null)
		setPhone(undefined)
		setEnterprise(null)
		setAddress(null)
		setSubject(null)
		setMessage(null)
		setNameValid(false)
		setIsLoading(false)
	}, [setName, setEmail, setPhone, setEnterprise, setAddress, setSubject, setMessage, setNameValid, setIsLoading])

	return { resetForm }
}

// Email sending functionality
const useEmailSending = (
	name: string | null,
	email: string | null,
	phone: string | undefined,
	enterprise: string | null,
	address: string | null,
	subject: string | null,
	message: string | null,
	text: Record<string, string | string[]> | undefined,
	setIsLoading: (loading: boolean) => void,
	resetForm: () => void,
) => {
	const sendEmail = useCallback((): void => {
		sendContactEmail({
			name, email, phone, enterprise, address, subject, message,
			text, setIsLoading, resetForm,
		})
	}, [name, email, phone, enterprise, address, subject, message, text, setIsLoading, resetForm])

	return { sendEmail }
}

// Validator creation functions
const useValidatorCreation = (
	name: string | null,
	email: string | null,
	message: string | null,
) => {
	const getNameValidators = useCallback((errorName: string): IValidator[] => [
		{
			validationFunction: () => validateName(name),
			errorMessage: errorName || 'Name is required',
		},
	], [name])

	const getEmailValidators = useCallback((errorEmail: string[]): IValidator[] => [
		{
			validationFunction: () => email !== null && email.trim() !== '',
			errorMessage: errorEmail[0] || 'Email is required',
		},
		{
			validationFunction: () => validateEmail(email),
			errorMessage: errorEmail[1] || 'Email format is invalid',
		},
	], [email])

	const getAddressValidators = useCallback((errorMessage: string): IValidator[] => [
		{
			validationFunction: () => validateMessage(message),
			errorMessage: errorMessage || 'Message is required',
		},
	], [message])

	return { getNameValidators, getEmailValidators, getAddressValidators }
}

// Form submission handling
const useFormSubmission = (checkInputs: () => boolean, sendEmail: () => void) => {
	const handleSubmit = useCallback((e: React.FormEvent) => {
		e.preventDefault()
		if (checkInputs()) {
			sendEmail()
		}
	}, [checkInputs, sendEmail])

	return { handleSubmit }
}

export const useContactFormValidation = (text?: Record<string, string | string[]>): IContactFormValidation => {
	const formState = useFormState()
	const formValidation = useFormValidation(formState.nameValid, formState.phone)
	const { resetForm } = useFormReset(
		formState.setName,
		formState.setEmail,
		formState.setPhone,
		formState.setEnterprise,
		formState.setAddress,
		formState.setSubject,
		formState.setMessage,
		formState.setNameValid,
		formState.setIsLoading,
	)
	const { sendEmail } = useEmailSending(
		formState.name,
		formState.email,
		formState.phone,
		formState.enterprise,
		formState.address,
		formState.subject,
		formState.message,
		text,
		formState.setIsLoading,
		resetForm,
	)
	const validatorCreation = useValidatorCreation(formState.name, formState.email, formState.message)
	const { handleSubmit } = useFormSubmission(formValidation.checkInputs, sendEmail)

	return {
		// Form field states
		name: formState.name,
		email: formState.email,
		phone: formState.phone,
		enterprise: formState.enterprise,
		address: formState.address,
		subject: formState.subject,
		message: formState.message,

		// Validation states
		nameValid: formState.nameValid,
		phoneError: null, // No longer tracking phone errors separately
		isLoading: formState.isLoading,

		// Form handlers
		setName: formState.setName,
		setEmail: formState.setEmail,
		setPhone: formState.setPhone,
		setEnterprise: formState.setEnterprise,
		setAddress: formState.setAddress,
		setSubject: formState.setSubject,
		setMessage: formState.setMessage,
		setNameValid: formState.setNameValid,
		setIsLoading: formState.setIsLoading,

		// Validation methods
		validatePhone: formValidation.validatePhone,
		validateForm: formValidation.validateForm,
		resetForm,
		handleSubmit,

		// Validator configurations
		getNameValidators: validatorCreation.getNameValidators,
		getEmailValidators: validatorCreation.getEmailValidators,
		getAddressValidators: validatorCreation.getAddressValidators,
	}
}

// Email sending helper function
const sendContactEmail = (params: {
	name: string | null, email: string | null, phone: string | undefined,
	enterprise: string | null, address: string | null, subject: string | null,
	message: string | null, text: Record<string, string | string[]> | undefined,
	setIsLoading: (loading: boolean) => void,
	resetForm: () => void
}) => {
	const { name, email, phone, enterprise, address, subject, message, text, setIsLoading, resetForm } = params
	const toastLoad = toast.dark(text?.sending)
	setIsLoading(true)
	emailjs
		.send(
			process.env.NEXT_PUBLIC_EMAILJS_SERVICE_ID ?? '',
			process.env.NEXT_PUBLIC_EMAILJS_TEMPLATE_ID ?? '',
			{
					text: message,
					name,
					email,
					phone,
					enterprise,
					address,
					subject,
				},
				process.env.NEXT_PUBLIC_EMAILJS_PUBLIC_KEY ?? '',
			)
			.then(() => {
				toast.dismiss(toastLoad)
				toast.success(text?.messageSent)
				resetForm()
			})
			.catch(() => {
				toast.dismiss(toastLoad)
				toast.error(text?.messageNotSent)
			})
			.finally(() => {
				setIsLoading(false)
			})
}
