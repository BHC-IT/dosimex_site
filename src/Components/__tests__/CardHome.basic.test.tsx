import { render, screen, fireEvent } from '@testing-library/react'
import { describe, it, expect, vi, beforeEach } from 'vitest'

import { useIsMobile } from '../../Hooks/useIsMobile'
import { useText } from '../../Hooks/useText'
import CardHome from '../CardHome'

// Mock Next.js router
const mockPush = vi.fn()
vi.mock('next/router', () => ({
	useRouter: vi.fn(() => ({
		push: mockPush,
		pathname: '/',
		query: {},
		asPath: '/',
	})),
}))

// Mock useIsMobile hook
vi.mock('../../Hooks/useIsMobile', () => ({
	useIsMobile: vi.fn(),
}))

// Mock useText hook
vi.mock('../../Hooks/useText', () => ({
	useText: vi.fn(),
}))

// Mock react-parallax-tilt
vi.mock('react-parallax-tilt', () => ({
	default: ({ children, ...props }: any) => (
		<div
			data-testid='tilt-wrapper'
			{...props}
		>
			{children}
		</div>
	),
}))

// Test data factories
const createMockCardProps = (overrides = {}) => ({
	icon: '/Images/software-icon.png',
	title: 'Software Solutions',
	content:
		'Advanced radiation dosimetry software for comprehensive dose management and analysis.',
	route: '/software',
	...overrides,
})

const createMockStyles = () => ({
	container: {
		display: 'flex',
		flexDirection: 'column',
		alignItems: 'center',
		padding: '20px',
		margin: '10px',
		borderRadius: '10px',
		backgroundColor: '#ffffff',
		boxShadow: '0 4px 8px rgba(0,0,0,0.1)',
		cursor: 'pointer',
		transition: 'all 0.3s ease',
	},
	image: {
		width: '80px',
		height: '80px',
		marginBottom: '15px',
	},
	title: {
		fontSize: '1.5rem',
		fontWeight: 'bold',
		color: '#333',
		marginBottom: '10px',
		textAlign: 'center',
	},
	content: {
		fontSize: '1rem',
		color: '#666',
		textAlign: 'center',
		lineHeight: 1.5,
	},
})

describe('CardHome Basic Functionality', () => {
	beforeEach(() => {
		vi.clearAllMocks()
		;(useIsMobile as any).mockReturnValue(createMockStyles())
		;(useText as any).mockReturnValue({
			altText: { cardIcon: 'Card icon' },
		})
	})

	describe('Rendering', () => {
		it('should render card with all content correctly', () => {
			const props = createMockCardProps()
			const { container } = render(<CardHome {...props} />)

			expect(container.firstChild).toMatchSnapshot()
		})

		it('should render card title', () => {
			const props = createMockCardProps()
			render(<CardHome {...props} />)

			expect(screen.getByText('Software Solutions')).toBeInTheDocument()
		})

		it('should render card content', () => {
			const props = createMockCardProps()
			render(<CardHome {...props} />)

			expect(
				screen.getByText(
					'Advanced radiation dosimetry software for comprehensive dose management and analysis.',
				),
			).toBeInTheDocument()
		})

		it('should render card with image', () => {
			const props = createMockCardProps()
			render(<CardHome {...props} />)

			const image = screen.getByRole('img')
			expect(image).toBeInTheDocument()
			expect(image).toHaveAttribute('alt', 'Card icon - Software Solutions')
		})

		it('should render tilt wrapper component', () => {
			const props = createMockCardProps()
			render(<CardHome {...props} />)

			expect(screen.getByTestId('tilt-wrapper')).toBeInTheDocument()
		})
	})

	describe('Props Handling', () => {
		it('should handle different icon paths', () => {
			const props = createMockCardProps({ icon: '/Images/training-icon.png' })
			render(<CardHome {...props} />)

			const image = screen.getByRole('img')
			expect(image).toHaveAttribute('src', expect.stringContaining('training-icon.png'))
		})

		it('should handle different titles', () => {
			const props = createMockCardProps({ title: 'Training Programs' })
			render(<CardHome {...props} />)

			expect(screen.getByText('Training Programs')).toBeInTheDocument()
		})

		it('should handle different content', () => {
			const props = createMockCardProps({
				content: 'Comprehensive training programs for professionals.',
			})
			render(<CardHome {...props} />)

			expect(
				screen.getByText('Comprehensive training programs for professionals.'),
			).toBeInTheDocument()
		})

		it('should handle different routes', () => {
			const props = createMockCardProps({ route: '/training' })
			render(<CardHome {...props} />)

			const button = screen.getByRole('button')
			fireEvent.click(button)

			expect(mockPush).toHaveBeenCalledWith('/training')
		})
	})
})
