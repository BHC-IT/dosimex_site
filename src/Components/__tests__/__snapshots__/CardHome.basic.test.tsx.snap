// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CardHome Basic Functionality > Rendering > should render card with all content correctly 1`] = `
<div
  data-testid="tilt-wrapper"
  scale="1.05"
  tiltmaxanglex="10"
  tiltmaxangley="10"
>
  <button
    aria-label="Navigate to Software Solutions section - Advanced radiation dosimetry software for comprehensive dose management and analysis."
    style="background: white;"
  >
    <div
      style="width: 80px; height: 80px; margin-bottom: 15px;"
    >
      <img
        alt="Card icon - Software Solutions"
        data-nimg="1"
        decoding="async"
        height="80"
        loading="lazy"
        src="/_next/image?url=%2FImages%2Fsoftware-icon.png&w=256&q=80"
        srcset="/_next/image?url=%2FImages%2Fsoftware-icon.png&w=96&q=80 1x, /_next/image?url=%2FImages%2Fsoftware-icon.png&w=256&q=80 2x"
        style="color: transparent;"
        width="80"
      />
    </div>
    <h4>
      Software Solutions
    </h4>
    <p>
      Advanced radiation dosimetry software for comprehensive dose management and analysis.
    </p>
    <div>
      <div>
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-right-long "
          data-icon="right-long"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M334.5 414c8.8 3.8 19 2 26-4.6l144-136c4.8-4.5 7.5-10.8 7.5-17.4s-2.7-12.9-7.5-17.4l-144-136c-7-6.6-17.2-8.4-26-4.6s-14.5 12.5-14.5 22l0 72L32 192c-17.7 0-32 14.3-32 32l0 64c0 17.7 14.3 32 32 32l288 0 0 72c0 9.6 5.7 18.2 14.5 22z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
  </button>
</div>
`;
